hostfactory/1.1/providerplugins/azurecc/scripts/src/cluster.py

```diff
@@ -13,28 +13,85 @@
         self.provider_config = provider_config
         self.logger = logger or logging.getLogger()
+        CC_CONFIG = {}
+        CC_CONFIG["url"] = self.provider_config.get("cyclecloud.config.web_server")
+        CC_CONFIG["username"] = self.provider_config.get("cyclecloud.config.username")
+        CC_CONFIG["password"] = self.provider_config.get("cyclecloud.config.password")
+        CC_CONFIG["cluster_name"] = self.cluster_name
+        self.node_mgr = new_node_manager(CC_CONFIG)
     
     def status(self):
         return self.get("/clusters/%s/status" % self.cluster_name)
+    
+    #old add nodes function
+    # def add_nodes(self, request):
+    #     self.logger.debug(request)
+    #     response_raw = self.post("/clusters/%s/nodes/create" % self.cluster_name, json=request)
+    #     try:
+    #         return json.loads(response_raw)
+    #     except:
+    #         raise RuntimeError("Could not parse response as json to create_nodes! '%s'" % response_raw)
+    
+    def add_nodes(self,request):
+        sku=request['sets'][0]['definition']['machineType']
+        selector = {"node.nodearray": request['sets'][0]['nodearray']}
+        if sku:
+          selector["node.vm_size"] = sku
+        self.logger.debug(request)
+        r=self.node_mgr.allocate(selector, node_count=request['sets'][0]['count'])
+        self.logger.debug("Request id in add nodes %s",request['requestId'])
 
-    def add_nodes(self, request):
-        response_raw = self.post("/clusters/%s/nodes/create" % self.cluster_name, json=request)
-        try:
-            return json.loads(response_raw)
-        except:
-            raise RuntimeError("Could not parse response as json to create_nodes! '%s'" % response_raw)
+        request_id_start = f"{request['requestId']}-start"
+        request_id_create = f"{request['requestId']}-create"
+        bootpup_resp=self.node_mgr.bootup(
+            request_id_start=request_id_start, request_id_create=request_id_create
+        )
+        self.logger.debug("node bootup %s",bootpup_resp)
+        self.logger.debug("node bootup requestids %s",bootpup_resp.request_ids)
+        return (r)
     
     def all_nodes(self):
         return self.get("/clusters/%s/nodes" % self.cluster_name)
-
+    # old nodes function
+    # def nodes(self, request_ids):
+    #     responses = {}
+    #     for request_id in request_ids:
+    #        # responses[request_id] = self.get("/clusters/%s/nodes" % self.cluster_name, request_id=f"{request_id}-start")
+    #         responses[request_id]=self.get("/clusters/%s/nodes" % self.cluster_name, request_id=f"{request_id}-create")
+    #         self.logger.debug(responses)
+    #     return responses
+    
     def nodes(self, request_ids):
         responses = {}
         for request_id in request_ids:
-            responses[request_id] = self.get("/clusters/%s/nodes" % self.cluster_name, request_id=request_id)
+          request_id_start = f"{request_id}-start"
```

Comment:
maybe this would be best to have a small inner function that does the get_nodes_by_request_id and all of the exception handling, and then a just call it once with each request_id

ie something like...
def _get_nodes_by_request_id(req_id, action):
    try:
        nodes_affected = self.node_mgr.get_nodes_by_request_id(req_id)
    except Exception as e:
        if "No operation found for request id" in str(e):
            self.logger.debug(f"No new nodes have been {action}")
        self.logger.debug(f"Nodes {action} %s", nodes_affected)
nodes_started = _get_nodes_by_request_id(request_id_started, "started")
nodes_created = _get_nodes_by_request_id(request_id_created, "created")
responses[request_id] = nodes_created + nodes_started

Reply:
yes, I can do it that way it makes sense